In [1]:

import pandas as pd

In [2]:
df = pd.read_csv('../data/Dataset.csv')
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,4.2,No,2-Day Shipping,No,No,32,Venmo,Weekly
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,4.5,No,Store Pickup,No,No,41,Bank Transfer,Bi-Weekly
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,2.9,No,Standard,No,No,24,Venmo,Quarterly
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,3.8,No,Express,No,No,24,Venmo,Weekly


In [3]:
df.isnull().sum()

Customer ID                0
Age                        0
Gender                     0
Item Purchased             0
Category                   0
Purchase Amount (USD)      0
Location                   0
Size                       0
Color                      0
Season                     0
Review Rating             37
Subscription Status        0
Shipping Type              0
Discount Applied           0
Promo Code Used            0
Previous Purchases         0
Payment Method             0
Frequency of Purchases     0
dtype: int64

In [4]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer ID             3900 non-null   int64  
 1   Age                     3900 non-null   int64  
 2   Gender                  3900 non-null   str    
 3   Item Purchased          3900 non-null   str    
 4   Category                3900 non-null   str    
 5   Purchase Amount (USD)   3900 non-null   int64  
 6   Location                3900 non-null   str    
 7   Size                    3900 non-null   str    
 8   Color                   3900 non-null   str    
 9   Season                  3900 non-null   str    
 10  Review Rating           3863 non-null   float64
 11  Subscription Status     3900 non-null   str    
 12  Shipping Type           3900 non-null   str    
 13  Discount Applied        3900 non-null   str    
 14  Promo Code Used         3900 non-null   str    
 15

In [5]:
df.duplicated().sum()

np.int64(0)

In [6]:
df.shape

(3900, 18)

# Lets First Identify the unique and nunique

> Customer Id Is Always unique

In [7]:
# Lets First Identify the unique and nunique
print(df['Age'].unique())
print(df['Age'].nunique())
print(df['Age'].value_counts())

[55 19 50 21 45 46 63 27 26 57 53 30 61 65 64 25 52 66 31 56 18 38 54 33
 36 35 29 70 69 67 20 39 42 68 49 59 47 40 41 48 22 24 44 37 58 32 62 51
 28 43 34 23 60]
53
Age
69    88
57    87
41    86
25    85
49    84
50    83
27    83
54    83
62    83
32    82
19    81
58    81
42    80
31    79
28    79
43    79
37    77
46    76
29    76
63    75
68    75
59    75
56    74
36    74
55    73
64    73
52    73
45    72
65    72
35    72
40    72
51    72
30    71
66    71
47    71
23    71
53    70
38    70
21    69
26    69
18    69
39    68
48    68
24    68
34    68
70    67
22    66
61    65
60    65
33    63
20    62
67    54
44    51
Name: count, dtype: int64


In [8]:
# 2 Male are more than Female

print(df['Gender'].unique())
print(df['Gender'].nunique())
print(df['Gender'].value_counts())

<StringArray>
['Male', 'Female']
Length: 2, dtype: str
2
Gender
Male      2652
Female    1248
Name: count, dtype: int64


In [9]:
# Item Purchased - Obj

print(df['Item Purchased'].unique())
print(df['Item Purchased'].nunique())
print(df['Item Purchased'].value_counts().sort_values(ascending = False).reset_index())

<StringArray>
[    'Blouse',    'Sweater',      'Jeans',    'Sandals',   'Sneakers',
      'Shirt',     'Shorts',       'Coat',    'Handbag',      'Shoes',
      'Dress',      'Skirt', 'Sunglasses',      'Pants',     'Jacket',
     'Hoodie',    'Jewelry',    'T-shirt',      'Scarf',        'Hat',
      'Socks',   'Backpack',       'Belt',      'Boots',     'Gloves']
Length: 25, dtype: str
25
   Item Purchased  count
0          Blouse    171
1           Pants    171
2         Jewelry    171
3           Shirt    169
4           Dress    166
5         Sweater    164
6          Jacket    163
7            Coat    161
8      Sunglasses    161
9            Belt    161
10        Sandals    160
11          Socks    159
12          Skirt    158
13         Shorts    157
14          Scarf    157
15            Hat    154
16        Handbag    153
17         Hoodie    151
18          Shoes    150
19        T-shirt    147
20       Sneakers    145
21          Boots    144
22       Backpack    143
23   

In [10]:

from rapidfuzz import process

# Step 1 - get your clean reference list (top frequent values)
reference = df['Item Purchased'].value_counts().head(30).index.tolist()

# Step 2 - map every GPU value to its closest match
def fuzzy_fix(val, ref, threshold=85):
    result = process.extractOne(val, ref)
    if result and result[1] >= threshold:
        return result[0]
    return val   # keep original if no good match found

df['Item Purchased'] = df['Item Purchased'].apply(lambda x: fuzzy_fix(x, reference))

# Step 3
print("After: ", df['Item Purchased'].nunique())

After:  25


In [11]:
# Category - Obj

print(df['Category'].unique())
print(df['Category'].nunique())
print(df['Category'].value_counts().sort_values(ascending = False).reset_index())

<StringArray>
['Clothing', 'Footwear', 'Outerwear', 'Accessories']
Length: 4, dtype: str
4
      Category  count
0     Clothing   1737
1  Accessories   1240
2     Footwear    599
3    Outerwear    324


In [12]:
# Purchased Amount(USD) - Obj

print(df['Purchase Amount (USD)'].unique())
print(df['Purchase Amount (USD)'].nunique())
print(df['Purchase Amount (USD)'].value_counts().sort_values(ascending = False).reset_index())

[ 53  64  73  90  49  20  85  34  97  31  68  72  51  81  36  38  48  62
  37  88  22  25  56  94  79  67  91  33  69  45  60  76  39 100  21  43
  54  55  30  28  59  24  83  46  70  29  26  58  96  40  32  95  41  47
  99  87  92  98  89  35  75  65  86  84  77  27  23  74  80  42  61  82
  78  93  50  57  66  44  52  71  63]
81
    Purchase Amount (USD)  count
0                      36     62
1                      94     62
2                      32     62
3                      51     61
4                      90     60
..                    ...    ...
76                     47     36
77                     49     35
78                     87     35
79                     69     34
80                     61     33

[81 rows x 2 columns]


In [13]:
# Location - Obj

print(df['Location'].unique())
print(df['Location'].nunique())
print(df['Location'].value_counts().sort_values(ascending = False).reset_index())

<StringArray>
[      'Kentucky',          'Maine',  'Massachusetts',   'Rhode Island',
         'Oregon',        'Wyoming',        'Montana',      'Louisiana',
  'West Virginia',       'Missouri',       'Arkansas',         'Hawaii',
       'Delaware',  'New Hampshire',       'New York',        'Alabama',
    'Mississippi', 'North Carolina',     'California',       'Oklahoma',
        'Florida',          'Texas',         'Nevada',         'Kansas',
       'Colorado',   'North Dakota',       'Illinois',        'Indiana',
        'Arizona',         'Alaska',      'Tennessee',           'Ohio',
     'New Jersey',       'Maryland',        'Vermont',     'New Mexico',
 'South Carolina',          'Idaho',   'Pennsylvania',    'Connecticut',
           'Utah',       'Virginia',        'Georgia',       'Nebraska',
           'Iowa',   'South Dakota',      'Minnesota',     'Washington',
      'Wisconsin',       'Michigan']
Length: 50, dtype: str
50
          Location  count
0          Montana   

In [14]:
# step 2 for rapidfuzz
# pip install rapidfuzz install this first
from rapidfuzz import process

# Step 1 - get your clean reference list (top frequent values)
reference = df['Location'].value_counts().head(30).index.tolist()

# Step 2 - map every GPU value to its closest match
def fuzzy_fix(val, ref, threshold=85):
    result = process.extractOne(val, ref)
    if result and result[1] >= threshold:
        return result[0]
    return val   # keep original if no good match found

df['Location'] = df['Location'].apply(lambda x: fuzzy_fix(x, reference))

# Step 3
print("After: ", df['Location'].nunique())

After:  49


In [15]:
# Size - Obj

print(df['Size'].unique())
print(df['Size'].nunique())
print(df['Size'].value_counts().sort_values(ascending = False).reset_index())

<StringArray>
['L', 'S', 'M', 'XL']
Length: 4, dtype: str
4
  Size  count
0    M   1755
1    L   1053
2    S    663
3   XL    429


In [16]:
# Color - Obj


print(df['Color'].unique())
print(df['Color'].nunique())
print(df['Color'].value_counts().sort_values(ascending = False).reset_index())

<StringArray>
[     'Gray',    'Maroon', 'Turquoise',     'White',  'Charcoal',    'Silver',
      'Pink',    'Purple',     'Olive',      'Gold',    'Violet',      'Teal',
  'Lavender',     'Black',     'Green',     'Peach',       'Red',      'Cyan',
     'Brown',     'Beige',    'Orange',    'Indigo',    'Yellow',   'Magenta',
      'Blue']
Length: 25, dtype: str
25
        Color  count
0       Olive    177
1      Yellow    174
2      Silver    173
3        Teal    172
4       Green    169
5       Black    167
6      Violet    166
7        Cyan    166
8        Gray    159
9      Maroon    158
10     Orange    154
11   Charcoal    153
12       Pink    153
13    Magenta    152
14       Blue    152
15     Purple    151
16      Peach    149
17        Red    148
18   Lavender    147
19      Beige    147
20     Indigo    147
21  Turquoise    145
22      White    142
23      Brown    141
24       Gold    138


In [17]:
# Season - Obj

print(df['Season'].unique())
print(df['Season'].nunique())
print(df['Season'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
['Winter', 'Spring', 'Summer', 'Fall']
Length: 4, dtype: str
4
   Season  count
0  Spring    999
1    Fall    975
2  Winter    971
3  Summer    955


In [18]:
# Review Rating - Obj

print(df['Review Rating'].unique())
print(df['Review Rating'].nunique())
print(df['Review Rating'].value_counts().sort_values(ascending = False).reset_index())


[3.1 3.5 2.7 2.9 3.2 2.6 4.8 4.1 4.9 4.5 4.7 2.8 4.6 3.3 4.4 3.6 5.  4.
 nan 4.2 3.7 3.9 3.  3.8 3.4 4.3 2.5]
26
    Review Rating  count
0             3.4    182
1             4.0    181
2             4.6    170
3             4.2    169
4             2.9    166
5             4.9    162
6             3.9    162
7             3.0    162
8             2.6    158
9             4.4    158
10            3.1    156
11            2.7    154
12            3.5    152
13            3.2    152
14            3.7    149
15            4.1    148
16            4.7    148
17            3.6    147
18            4.3    147
19            3.3    146
20            4.8    144
21            3.8    141
22            4.5    139
23            2.8    136
24            5.0     68
25            2.5     66


In [19]:
# Subscription Status - Obj

print(df['Subscription Status'].unique())
print(df['Subscription Status'].nunique())
print(df['Subscription Status'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
['Yes', 'No']
Length: 2, dtype: str
2
  Subscription Status  count
0                  No   2847
1                 Yes   1053


In [20]:
# Shipping Type - Obj

print(df['Shipping Type'].unique())
print(df['Shipping Type'].nunique())
print(df['Shipping Type'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
[       'Express',  'Free Shipping',   'Next Day Air',       'Standard',
 '2-Day Shipping',   'Store Pickup']
Length: 6, dtype: str
6
    Shipping Type  count
0   Free Shipping    675
1        Standard    654
2    Store Pickup    650
3    Next Day Air    648
4         Express    646
5  2-Day Shipping    627


In [21]:
# Discount Applied - Obj

print(df['Discount Applied'].unique())
print(df['Discount Applied'].nunique())
print(df['Discount Applied'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
['Yes', 'No']
Length: 2, dtype: str
2
  Discount Applied  count
0               No   2223
1              Yes   1677


In [22]:
# Promo Code Used - Obj

print(df['Promo Code Used'].unique())
print(df['Promo Code Used'].nunique())
print(df['Promo Code Used'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
['Yes', 'No']
Length: 2, dtype: str
2
  Promo Code Used  count
0              No   2223
1             Yes   1677


In [23]:
# Previous Purchases - Obj

print(df['Previous Purchases'].unique())
print(df['Previous Purchases'].nunique())
print(df['Previous Purchases'].value_counts().sort_values(ascending = False).reset_index())


[14  2 23 49 31 19  8  4 26 10 37 34 44 36 17 46 50 22 32 40 16 13  7 41
 45 38 48 18 15 25 39 35 29 21 43  3  5 24 42 47 28 20 33  1  9 12 27 11
 30  6]
50
    Previous Purchases  count
0                   31     97
1                   21     96
2                   14     92
3                    4     91
4                    3     91
5                   24     91
6                   48     90
7                   47     90
8                    5     87
9                    6     87
10                  18     84
11                  45     83
12                  42     83
13                  20     83
14                   1     83
15                  40     82
16                  33     82
17                  11     82
18                  30     82
19                  32     81
20                  39     81
21                  28     80
22                  34     79
23                  25     79
24                  12     79
25                  19     78
26                  37     78
27 

In [24]:
# Payment Method - Obj

print(df['Payment Method'].unique())
print(df['Payment Method'].nunique())
print(df['Payment Method'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
['Venmo', 'Cash', 'Credit Card', 'PayPal', 'Bank Transfer', 'Debit Card']
Length: 6, dtype: str
6
  Payment Method  count
0         PayPal    677
1    Credit Card    671
2           Cash    670
3     Debit Card    636
4          Venmo    634
5  Bank Transfer    612


In [25]:
# Frequency of Purchases - Obj

print(df['Frequency of Purchases'].unique())
print(df['Frequency of Purchases'].nunique())
print(df['Frequency of Purchases'].value_counts().sort_values(ascending = False).reset_index())


<StringArray>
[   'Fortnightly',         'Weekly',       'Annually',      'Quarterly',
      'Bi-Weekly',        'Monthly', 'Every 3 Months']
Length: 7, dtype: str
7
  Frequency of Purchases  count
0         Every 3 Months    584
1               Annually    572
2              Quarterly    563
3                Monthly    553
4              Bi-Weekly    547
5            Fortnightly    542
6                 Weekly    539


In [26]:
print('Category_Columns:', df.select_dtypes(include = 'object').columns.to_list())
print('___________________________________________________________________________________________________________________________')
print("Numerical Columns:",df.select_dtypes(include = 'number').columns.tolist())

Category_Columns: ['Gender', 'Item Purchased', 'Category', 'Location', 'Size', 'Color', 'Season', 'Subscription Status', 'Shipping Type', 'Discount Applied', 'Promo Code Used', 'Payment Method', 'Frequency of Purchases']
___________________________________________________________________________________________________________________________
Numerical Columns: ['Customer ID', 'Age', 'Purchase Amount (USD)', 'Review Rating', 'Previous Purchases']


C:\Users\User\AppData\Local\Temp\ipykernel_27516\1533446027.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print('Category_Columns:', df.select_dtypes(include = 'object').columns.to_list())


In [27]:
print('-------------------------3.1 Inconsistent Spacing & Tab Spacing ----------------------------------')
cat_cols = ['Gender', 'Item Purchased', 'Category',
            'Location', 'Size', 'Color', 'Season',
            'Subscription Status', 'Shipping Type',
            'Discount Applied', 'Promo Code Used',
            'Payment Method', 'Frequency of Purchases']

for col in cat_cols:
    before = df[col].nunique()

    if df[col].dtype == 'object':
        # string column — apply spacing fix
        df[col] = df[col].str.replace(r'\s+', ' ', regex=True).str.strip()
    else:
        # int or float column — no spacing possible, skip
        pass

    after = df[col].nunique()
    print(f"{col} | dtype: {df[col].dtype} | before: {before} | after: {after}")

-------------------------3.1 Inconsistent Spacing & Tab Spacing ----------------------------------
Gender | dtype: str | before: 2 | after: 2
Item Purchased | dtype: str | before: 25 | after: 25
Category | dtype: str | before: 4 | after: 4
Location | dtype: str | before: 49 | after: 49
Size | dtype: str | before: 4 | after: 4
Color | dtype: str | before: 25 | after: 25
Season | dtype: str | before: 4 | after: 4
Subscription Status | dtype: str | before: 2 | after: 2
Shipping Type | dtype: str | before: 6 | after: 6
Discount Applied | dtype: str | before: 2 | after: 2
Promo Code Used | dtype: str | before: 2 | after: 2
Payment Method | dtype: str | before: 6 | after: 6
Frequency of Purchases | dtype: str | before: 7 | after: 7


In [28]:
df['Review Rating']= df['Review Rating'].fillna(df['Review Rating'].mode()[0])

In [29]:
pip install rapidfuzz

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


# i have checked the data in depth
Here are the features you should engineer with their logic:

> Case - 1

# Lets understand the data for now.
# Feature 1 — discount_dependency_flag

In [30]:
print(pd.crosstab(df['Discount Applied'], df['Promo Code Used']))

Promo Code Used     No   Yes
Discount Applied            
No                2223     0
Yes                  0  1677


In [31]:
print(df['Discount Applied'].unique())

<StringArray>
['Yes', 'No']
Length: 2, dtype: str


In [32]:
# Check the case YOU were worried about
case1 = df[(df['Discount Applied'] == 'Yes') & (df['Promo Code Used'] == 'No')]
print('Discount Yes but Promo No:', len(case1))

# Check the opposite case too
case2 = df[(df['Discount Applied'] == 'No') & (df['Promo Code Used'] == 'Yes')]
print('Discount No but Promo Yes:', len(case2))

# check equal concept len of YES
case3 = df[(df['Discount Applied'] == 'Yes') & (df['Promo Code Used'] == 'Yes')]
print('Discount No but Promo Yes:', len(case3))

# check equal concept len NO
case4 = df[(df['Discount Applied'] == 'No') & (df['Promo Code Used'] == 'No')]
print('Discount No but Promo NO:', len(case4))

Discount Yes but Promo No: 0
Discount No but Promo Yes: 0
Discount No but Promo Yes: 1677
Discount No but Promo NO: 2223


In [33]:

df['Discount dependency Flag'] = (
    (df['Discount Applied'] == 'Yes') &
    (df['Promo Code Used'] == 'Yes')
).astype(int)
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly,1
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly,1
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly,1
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly,1
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,4.2,No,2-Day Shipping,No,No,32,Venmo,Weekly,0
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,4.5,No,Store Pickup,No,No,41,Bank Transfer,Bi-Weekly,0
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,2.9,No,Standard,No,No,24,Venmo,Quarterly,0
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,3.8,No,Express,No,No,24,Venmo,Weekly,0


> Logic: A customer who ALWAYS needs both a discount AND a promo code to buy is not loyal — they are a deal hunter. This flag identifies them. If this is 1, the brand is paying to acquire that purchase, not earning it.

In [34]:
# Feature 2 — frequency_score
df['Frequency of Purchases'].value_counts()

Frequency of Purchases
Every 3 Months    584
Annually          572
Quarterly         563
Monthly           553
Bi-Weekly         547
Fortnightly       542
Weekly            539
Name: count, dtype: int64

In [35]:
# Step 2:

In [36]:
frequency_map = {
    'Weekly': 7,
    'Fortnightly': 6,
    'Bi-Weekly': 5,
    'Monthly': 4,
    'Every 3 Months': 3,
    'Quarterly': 2,
    'Annually': 1
}

df['Frequency Score']= df['Frequency of Purchases'].map(frequency_map)
df
# number increases purchases increases more valuable

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,Review Rating,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,3.1,Yes,Express,Yes,Yes,14,Venmo,Fortnightly,1,6
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,3.1,Yes,Express,Yes,Yes,2,Cash,Fortnightly,1,6
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,3.1,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly,1,7
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,3.5,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly,1,7
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,2.7,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,4.2,No,2-Day Shipping,No,No,32,Venmo,Weekly,0,7
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,4.5,No,Store Pickup,No,No,41,Bank Transfer,Bi-Weekly,0,5
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,2.9,No,Standard,No,No,24,Venmo,Quarterly,0,2
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,3.8,No,Express,No,No,24,Venmo,Weekly,0,7


> Logic: Frequency of Purchases is text — Python cannot compare "Weekly" vs "Annually" mathematically. This converts it into a number so you can rank, average, and segment customers by how often they buy.

In [37]:
# Feature 3 — customer_value_score
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Customer ID               3900 non-null   int64  
 1   Age                       3900 non-null   int64  
 2   Gender                    3900 non-null   str    
 3   Item Purchased            3900 non-null   str    
 4   Category                  3900 non-null   str    
 5   Purchase Amount (USD)     3900 non-null   int64  
 6   Location                  3900 non-null   str    
 7   Size                      3900 non-null   str    
 8   Color                     3900 non-null   str    
 9   Season                    3900 non-null   str    
 10  Review Rating             3900 non-null   float64
 11  Subscription Status       3900 non-null   str    
 12  Shipping Type             3900 non-null   str    
 13  Discount Applied          3900 non-null   str    
 14  Promo Code Used    

# Step 3:

In [38]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3900 entries, 0 to 3899
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Customer ID               3900 non-null   int64  
 1   Age                       3900 non-null   int64  
 2   Gender                    3900 non-null   str    
 3   Item Purchased            3900 non-null   str    
 4   Category                  3900 non-null   str    
 5   Purchase Amount (USD)     3900 non-null   int64  
 6   Location                  3900 non-null   str    
 7   Size                      3900 non-null   str    
 8   Color                     3900 non-null   str    
 9   Season                    3900 non-null   str    
 10  Review Rating             3900 non-null   float64
 11  Subscription Status       3900 non-null   str    
 12  Shipping Type             3900 non-null   str    
 13  Discount Applied          3900 non-null   str    
 14  Promo Code Used    

In [39]:
import pandas as pd

# Creating the DataFrame
data = {
    "Component": [
        "Purchase Amount",
        "Frequency Score",
        "Previous Purchases"
    ],
    "What It Adds to Score": [
        "How much per visit",
        "How often they visit",
        "How long they have been visiting"
    ]
}

gf = pd.DataFrame(data)

> Justifying why multipy not addition
-

# Customer Value Score Calculation

The **Value Score** is computed as:

`Value Score = Spend × Frequency Score × Previous Purchases`

Where:
- **Spend** = average transaction amount ($)
- **Frequency Score** = number of purchases per period (e.g., weekly = 7, quarterly = 2)
- **Previous Purchases** = total historical purchase count

| Customer | Spend ($) | Frequency Score | Previous Purchases | Value Score |
|----------|-----------|----------------|--------------------|--------------|
| A        | 80        | 7 (Weekly)      | 40                 | 80 × 7 × 40 = **22,400** |
| B        | 80        | 2 (Quarterly)   | 40                 | 80 × 2 × 40 = **6,400** |
| C        | 80        | 7 (Weekly)      | 5                  | 80 × 7 × 5 = **2,800** |

- **Customer A** – high frequency + high recency → highest value.
- **Customer B** – low frequency reduces value despite high past purchases.
- **Customer C** – low historical purchases limits total value.

In [40]:
# It is a single number that represents how much total value a customer could generate for the brand over time.

df['Customer Value Score'] = (
    df['Purchase Amount (USD)'] * df['Frequency Score'] * df['Previous Purchases']
)
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Subscription Status,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Yes,Express,Yes,Yes,14,Venmo,Fortnightly,1,6,4452
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Yes,Express,Yes,Yes,2,Cash,Fortnightly,1,6,768
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Yes,Free Shipping,Yes,Yes,23,Credit Card,Weekly,1,7,11753
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Yes,Next Day Air,Yes,Yes,49,PayPal,Weekly,1,7,30870
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Yes,Free Shipping,Yes,Yes,31,PayPal,Annually,1,1,1519
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,No,2-Day Shipping,No,No,32,Venmo,Weekly,0,7,6272
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,No,Store Pickup,No,No,41,Bank Transfer,Bi-Weekly,0,5,10045
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,No,Standard,No,No,24,Venmo,Quarterly,0,2,1584
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,No,Express,No,No,24,Venmo,Weekly,0,7,12936


> Logic: A customer who spends $200 once a year is less valuable than one who spends $80 every week. This score combines all three dimensions — how much, how often, and how many times before — into one number that represents true lifetime value potential.

In [41]:
# Step 4:

In [42]:
df['Organic Buyer Flag'] = (
    (df['Discount Applied'] == 'No') &
    (df['Promo Code Used'] == 'No')
).astype(int)

> Logic: This customer bought at full price with no incentive. That is the most valuable type of customer — they chose the brand on its own merit. The brand does NOT need to spend money to bring them back.

In [43]:
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Shipping Type,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Express,Yes,Yes,14,Venmo,Fortnightly,1,6,4452,0
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Express,Yes,Yes,2,Cash,Fortnightly,1,6,768,0
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Free Shipping,Yes,Yes,23,Credit Card,Weekly,1,7,11753,0
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Next Day Air,Yes,Yes,49,PayPal,Weekly,1,7,30870,0
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Free Shipping,Yes,Yes,31,PayPal,Annually,1,1,1519,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,2-Day Shipping,No,No,32,Venmo,Weekly,0,7,6272,1
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,Store Pickup,No,No,41,Bank Transfer,Bi-Weekly,0,5,10045,1
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,Standard,No,No,24,Venmo,Quarterly,0,2,1584,1
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,Express,No,No,24,Venmo,Weekly,0,7,12936,1


In [44]:
df.groupby('Organic Buyer Flag')['Purchase Amount (USD)'].mean()

Organic Buyer Flag
0    59.279070
1    60.130454
Name: Purchase Amount (USD), dtype: float64

> High spend + discount dependency = dangerous customer. The moment you stop discounting, they leave.

> High frequency + low discount = safe customer. They stay regardless.

> Loyalty Definition A → Built on frequency + history + no discount
                     → These customers may spend LESS per transaction

> Loyalty Definition B → Built on subscription + spend + rating
                     → These customers DO spend more per transaction

In [45]:
# Score A Behavioral loyalty — frequency + history + no discount
df['Loyalty Score A'] = (
    df['Frequency Score']
    + df['Previous Purchases']
    + df['Organic Buyer Flag']
)
df
# frequency score: # number increases purchases increases more valuable
# previous purchase already in numbers in count total
# Organic buyer Flag:: discount promo no used.

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Discount Applied,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag,Loyalty Score A
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Yes,Yes,14,Venmo,Fortnightly,1,6,4452,0,20
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Yes,Yes,2,Cash,Fortnightly,1,6,768,0,8
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Yes,Yes,23,Credit Card,Weekly,1,7,11753,0,30
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Yes,Yes,49,PayPal,Weekly,1,7,30870,0,56
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Yes,Yes,31,PayPal,Annually,1,1,1519,0,32
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,No,No,32,Venmo,Weekly,0,7,6272,1,40
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,No,No,41,Bank Transfer,Bi-Weekly,0,5,10045,1,47
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,No,No,24,Venmo,Quarterly,0,2,1584,1,27
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,No,No,24,Venmo,Weekly,0,7,12936,1,32


In [46]:
# Score B Commitment loyalty — subscription + rating + spend
df['Loyalty Score B'] = (
    (df['Subscription Status'] == 'Yes').astype(int)  +
    (df['Review Rating'] >= 4.0).astype(int)  +
    (df['Purchase Amount (USD)'] > df['Purchase Amount (USD)'].median()).astype(int)
)
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag,Loyalty Score A,Loyalty Score B
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Yes,14,Venmo,Fortnightly,1,6,4452,0,20,1
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Yes,2,Cash,Fortnightly,1,6,768,0,8,2
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Yes,23,Credit Card,Weekly,1,7,11753,0,30,2
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Yes,49,PayPal,Weekly,1,7,30870,0,56,2
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Yes,31,PayPal,Annually,1,1,1519,0,32,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,No,32,Venmo,Weekly,0,7,6272,1,40,1
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,No,41,Bank Transfer,Bi-Weekly,0,5,10045,1,47,1
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,No,24,Venmo,Quarterly,0,2,1584,1,27,0
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,No,24,Venmo,Weekly,0,7,12936,1,32,1


In [47]:
df.isnull().sum()

Customer ID                 0
Age                         0
Gender                      0
Item Purchased              0
Category                    0
Purchase Amount (USD)       0
Location                    0
Size                        0
Color                       0
Season                      0
Review Rating               0
Subscription Status         0
Shipping Type               0
Discount Applied            0
Promo Code Used             0
Previous Purchases          0
Payment Method              0
Frequency of Purchases      0
Discount dependency Flag    0
Frequency Score             0
Customer Value Score        0
Organic Buyer Flag          0
Loyalty Score A             0
Loyalty Score B             0
dtype: int64

In [48]:
print('Score A:', df['Loyalty Score A'].value_counts().sort_index())
print('Score B:', df['Loyalty Score B'].value_counts().sort_index())
# Higher number = More loyal.

Score A: Loyalty Score A
2      2
3     11
4     35
5     36
6     64
7     58
8     81
9     89
10    81
11    72
12    83
13    79
14    67
15    64
16    65
17    99
18    67
19    81
20    81
21    64
22    75
23    93
24    90
25    85
26    61
27    85
28    79
29    70
30    82
31    68
32    88
33    85
34    82
35    70
36    74
37    72
38    87
39    87
40    76
41    67
42    86
43    74
44    90
45    76
46    70
47    62
48    75
49    85
50    73
51    88
52    71
53    66
54    40
55    46
56    22
57    14
58     7
Name: count, dtype: int64
Score B: Loyalty Score B
0     866
1    1694
2    1113
3     227
Name: count, dtype: int64


- Now As we have built the understanding do the real  question thinking:


In [49]:
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag,Loyalty Score A,Loyalty Score B
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Yes,14,Venmo,Fortnightly,1,6,4452,0,20,1
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Yes,2,Cash,Fortnightly,1,6,768,0,8,2
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Yes,23,Credit Card,Weekly,1,7,11753,0,30,2
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Yes,49,PayPal,Weekly,1,7,30870,0,56,2
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Yes,31,PayPal,Annually,1,1,1519,0,32,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,No,32,Venmo,Weekly,0,7,6272,1,40,1
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,No,41,Bank Transfer,Bi-Weekly,0,5,10045,1,47,1
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,No,24,Venmo,Quarterly,0,2,1584,1,27,0
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,No,24,Venmo,Weekly,0,7,12936,1,32,1


In [50]:
# Average Loyalty Score A for discount dependent vs not
# 1.Who are the genuinely loyal customers vs. those who only buy when there is a discount?


print(df.groupby('Discount dependency Flag')['Loyalty Score A'].mean().round(2))
# Discount applied and promo code used -- discount dependency flag
# #df['Loyalty Score A'] = (df['Frequency Score'] + df['Previous Purchases'] + df['Organic Buyer Flag']
# frequency score: # number increases purchases increases more valuable
# previous purchase already in numbers in count total
# Organic buyer Flag:: discount promo no used.

Discount dependency Flag
0    29.99
1    29.72
Name: Loyalty Score A, dtype: float64


# Discount Dependency Flag
- 0    higher score    ← organic buyers should be more loyal
- 1    lower score     ← discount buyers should be less loyal

# What This Means For Your Project

> Loyal and discount-dependent customers score almost identically on behavioral loyalty. This means discount dependency alone does not separate loyal from non-loyal customers in this dataset.

#This Is Actually A Strong Insight

> "Loyalty Score A reveals no meaningful behavioral difference between discount-dependent and organic buyers — the gap is only 0.27 points on a 58-point scale. This confirms that discount usage alone cannot identify genuine loyalty. Deeper segmentation using subscription status, frequency, and purchase history is required."


# What This Tells You About Your Data
This finding connects directly back to what we found earlier — spend difference was only $0.86. Now loyalty score difference is only 0.27. The data is consistently saying:
-
> Discount hunters and loyal customers look identical on surface metrics. You need Score B and behavioral patterns to separate them.


In [51]:
import pandas as pd

data = {
    "Column": [
        "Previous Purchases",
        "Purchase Amount (USD)",
        "Loyalty Score A",
        "Loyalty Score B"
    ],
    "Why It Indicates Value": [
        "More purchases = proven long term value",
        "Higher spend = more revenue per visit",
        "Higher behavioral loyalty = longer relationship",
        "Higher commitment = more stable customer"
    ]
}

gf = pd.DataFrame(data)
print(gf)

                  Column                           Why It Indicates Value
0     Previous Purchases          More purchases = proven long term value
1  Purchase Amount (USD)            Higher spend = more revenue per visit
2        Loyalty Score A  Higher behavioral loyalty = longer relationship
3        Loyalty Score B         Higher commitment = more stable customer


In [52]:
# 2.What behavioral patterns today predict high customer value over time?
df


,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Promo Code Used,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag,Loyalty Score A,Loyalty Score B
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,Yes,14,Venmo,Fortnightly,1,6,4452,0,20,1
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,Yes,2,Cash,Fortnightly,1,6,768,0,8,2
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,Yes,23,Credit Card,Weekly,1,7,11753,0,30,2
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,Yes,49,PayPal,Weekly,1,7,30870,0,56,2
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,Yes,31,PayPal,Annually,1,1,1519,0,32,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,No,32,Venmo,Weekly,0,7,6272,1,40,1
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,No,41,Bank Transfer,Bi-Weekly,0,5,10045,1,47,1
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,No,24,Venmo,Quarterly,0,2,1584,1,27,0
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,No,24,Venmo,Weekly,0,7,12936,1,32,1


In [53]:
print(df[['Frequency Score',
          'Previous Purchases',
          'Purchase Amount (USD)',
          'Loyalty Score A',
          'Loyalty Score B',
          'Organic Buyer Flag',
          'Discount dependency Flag'
         ]].corrwith(df['Customer Value Score']).round(3).sort_values(ascending=False))

Loyalty Score A             0.666
Previous Purchases          0.601
Frequency Score             0.521
Purchase Amount (USD)       0.408
Loyalty Score B             0.241
Discount dependency Flag    0.016
Organic Buyer Flag         -0.016
dtype: float64


# `df.corr()` vs `df.corrwith()` – Quick Reference

| Feature               | `.corr()`                              | `.corrwith()`                              |
|-----------------------|----------------------------------------|---------------------------------------------|
| **How many targets**  | All columns vs all columns             | One specific column vs all others          |
| **Comparison**        | Every column vs every column           | Every column **vs one specific column**    |
| **Output**            | Full correlation matrix (n x n)        | Single Series (one value per column)       |
| **When to use**       | Exploring **all** pairwise relationships | You have **one target** column of interest |

# Correlation with Customer Value Score – Plain English

| Column | Correlation | Plain English |
|--------|-------------|----------------|
| Loyalty Score A | 0.666 | Strongest predictor — behavioral loyalty directly drives value |
| Previous Purchases | 0.601 | More purchase history = more valuable customer |
| Frequency Score | 0.521 | Buying more frequently = more valuable |
| Purchase Amount (USD) | 0.408 | Higher spend contributes but is not the strongest signal |
| Loyalty Score B | 0.241 | Commitment loyalty has weak link to value |
| Discount Dependency Flag | 0.016 | Almost zero — discount usage has no link to value |
| Organic Buyer Flag | -0.016 | Almost zero — buying without discount alone does not predict value |

# The 3 Key Findings

## Finding 1 — Behavioral Loyalty Predicts Value Most
Loyalty Score A = 0.666

A customer who buys frequently + has deep history + is organic = highest value predictor. This validates your entire Score A construction.

## Finding 2 — History Depth Matters More Than Spend
Previous Purchases = 0.601  
Purchase Amount    = 0.408

A customer who has bought 40 times at $50 is more valuable than a customer who bought 5 times at $90. Relationship depth beats transaction size.

## Finding 3 — Discount Usage Has Zero Relationship With Value
Discount Dependency Flag = 0.016  
Organic Buyer Flag       = -0.016

Whether a customer uses discounts or not has absolutely no impact on their value score. This is your strongest Q1 and Q2 combined insight.

## Your Q2 Answer — One Paragraph For Report

> "Correlation analysis reveals that behavioral loyalty score, purchase history depth, and buying frequency are the three strongest predictors of customer value — with correlations of 0.666, 0.601, and 0.521 respectively. Purchase amount, while contributing positively at 0.408, is not the primary value driver. Most critically, discount dependency shows near-zero correlation with customer value (0.016), confirming that promotional activity neither builds nor destroys customer value — it is behaviorally irrelevant to long-term worth."

# Q2 Summary Table

| Rank | Column | Correlation | Action For Brand |
|------|--------|-------------|------------------|
| 1 | Loyalty Score A | 0.666 | Protect and retain high Score A customers |
| 2 | Previous Purchases | 0.601 | Reward customers with deep history |
| 3 | Frequency Score | 0.521 | Incentivize frequent buying — not discounts |
| 4 | Purchase Amount | 0.408 | Secondary signal only |
| 5 | Discount Flag | 0.016 | Stop using discounts as value-building tool |

# **3.Which geographies and demographics are commercially underlevered?**
> Underlevered = places or people where the brand has not put enough effort — even though genuine demand exists there.
# Real-Life Example: Chai Seller in a City

| Area | People Who Buy | Discount Needed | Your Marketing Effort |
|------|----------------|----------------|------------------------|
| A    | 500            | Always need discount | High effort |
| B    | 100            | Never need discount — buy anyway | Zero effort |

## Comment / Business Insight

> Area B is underlevered. People there genuinely want your chai without being pushed.
> But you are ignoring them and spending all your money on Area A where people only buy when bribed

In [54]:
#Average Customer Value by Location

location_value = df.groupby('Location')['Customer Value Score'].mean().sort_values(ascending=False)
print(location_value)


Location
Pennsylvania      8293.540541
Arizona           7888.615385
Alaska            7829.625000
Hawaii            7687.092308
Tennessee         7293.792208
Alabama           7257.235955
Illinois          7185.521739
Wyoming           6758.661972
Georgia           6661.303797
Nevada            6658.471264
Iowa              6619.420290
Ohio              6535.922078
North Carolina    6522.756410
Minnesota         6447.454545
Oregon            6446.364865
New Mexico        6298.185185
Indiana           6183.037975
South Dakota      6096.328571
Utah              6002.295775
Rhode Island      5996.920635
Nebraska          5955.379310
Washington        5927.287671
New York          5924.835443
Arkansas          5920.075949
Mississippi       5917.625000
Louisiana         5889.023810
Maryland          5887.895349
Massachusetts     5815.694444
Virginia          5804.168831
Kentucky          5719.810127
North Dakota      5650.867470
Colorado          5612.813333
West Virginia     5561.358025
M

In [55]:
#Average Loyalty by Location

location_loyalty = df.groupby('Location')['Loyalty Score A'].mean().sort_values(ascending=False)
print(location_loyalty)

Location
Hawaii            34.046154
Alaska            32.986111
Arizona           32.923077
Wyoming           32.788732
Missouri          32.283951
Alabama           32.247191
Iowa              32.115942
Pennsylvania      32.081081
Illinois          31.500000
Arkansas          31.455696
Utah              31.422535
Maryland          31.093023
Kentucky          31.088608
Michigan          31.082192
Minnesota         30.977273
Tennessee         30.766234
New Mexico        30.567901
Mississippi       30.537500
Nevada            30.471264
Indiana           30.354430
Florida           30.338235
Oregon            30.216216
New York          30.037975
South Carolina    29.855263
Georgia           29.835443
Ohio              29.779221
Montana           29.677083
North Carolina    29.512821
Nebraska          29.287356
Delaware          29.093023
South Dakota      29.028571
Washington        29.027397
Vermont           28.870588
California        28.863158
Idaho             28.655914
Colorado   

# Organic Buyer Flag: no promo code no discount
# Customer Value Score: Value Score = Spend × Frequency Score × Previous Purchases`


In [56]:
#Organic Buyer Percentage by Location

organic_location = df.groupby('Location')['Organic Buyer Flag'].mean().sort_values(ascending=False)
print(organic_location)

Location
Kansas            0.761905
Connecticut       0.666667
Arizona           0.661538
Maine             0.649351
Tennessee         0.636364
Texas             0.636364
South Dakota      0.628571
Montana           0.625000
Virginia          0.623377
Vermont           0.611765
Rhode Island      0.603175
Michigan          0.602740
Illinois          0.597826
Alaska            0.597222
New Jersey        0.597015
Alabama           0.595506
Georgia           0.594937
Idaho             0.591398
Colorado          0.586667
Louisiana         0.583333
California        0.578947
Wyoming           0.577465
New York          0.575949
Nebraska          0.574713
Kentucky          0.569620
Washington        0.561644
Ohio              0.558442
Maryland          0.558140
Minnesota         0.556818
New Mexico        0.555556
Pennsylvania      0.554054
North Carolina    0.551282
Delaware          0.546512
Florida           0.544118
North Dakota      0.542169
Utah              0.535211
Arkansas          0

In [57]:
#Combined Table

geo_analysis = pd.DataFrame({
    'Avg Customer Value': location_value,
    'Avg Loyalty': location_loyalty,
    'Organic Buyer %': organic_location
})

print(geo_analysis)

                Avg Customer Value  Avg Loyalty  Organic Buyer %
Location                                                        
Alabama                7257.235955    32.247191         0.595506
Alaska                 7829.625000    32.986111         0.597222
Arizona                7888.615385    32.923077         0.661538
Arkansas               5920.075949    31.455696         0.531646
California             5390.294737    28.863158         0.578947
Colorado               5612.813333    28.440000         0.586667
Connecticut            4753.423077    28.320513         0.666667
Delaware               4919.872093    29.093023         0.546512
Florida                5475.691176    30.338235         0.544118
Georgia                6661.303797    29.835443         0.594937
Hawaii                 7687.092308    34.046154         0.507692
Idaho                  5355.290323    28.655914         0.591398
Illinois               7185.521739    31.500000         0.597826
Indiana                61

In [58]:
#Age Group Creation

bins = [18,25,35,45,55,70]

labels = ['18-25','26-35','36-45','46-55','56+']

df['Age Group'] = pd.cut(
    df['Age'],
    bins=bins,
    labels=labels
)

In [59]:
#Analyse Age Group Values

age_analysis = df.groupby('Age Group')['Customer Value Score'].mean().sort_values(ascending=False)
print(age_analysis)

Age Group
46-55    6236.598938
56+      6112.719457
26-35    5934.688679
36-45    5829.303155
18-25    5792.362550
Name: Customer Value Score, dtype: float64


In [60]:
#Age Group Organic Behaviour

age_organic = df.groupby('Age Group')['Organic Buyer Flag'].mean().sort_values(ascending=False)
print(age_organic)

Age Group
36-45    0.585734
46-55    0.567065
56+      0.566516
18-25    0.565737
26-35    0.564690
Name: Organic Buyer Flag, dtype: float64


Q.3 *INTERPRETATION*

In [61]:
print(df.groupby('Age Group')['Customer ID'].count())

Age Group
18-25     502
26-35     742
36-45     729
46-55     753
56+      1105
Name: Customer ID, dtype: int64


# Analysis: worth to know:
### Age Group vs Customer Value Analysis

```python
age_analysis = df.groupby('Age Group')['Customer Value Score'].mean().sort_values(ascending=False)
```

This analysis was performed to understand how customer value differs across various age demographics.

The dataset was grouped by Age Group, and the average Customer Value Score was calculated for each category. The results indicate that customers aged 36–45 exhibit the highest average customer value within the dataset.

This suggests that the 36–45 demographic contributes comparatively higher business value on average. A possible explanation could be that individuals within this age range are more financially stable, have established purchasing habits, or make family-oriented purchasing decisions.

However, this observation should not be interpreted as a universal behavioral rule. The analysis only identifies patterns present within the current dataset and does not establish direct causation between age and purchasing behavior.

Customer value can also be influenced by multiple additional factors such as:

* income level,
* family size,
* marital status,
* purchasing responsibility,
* occupation,
* lifestyle preferences.

Therefore, age should be interpreted as one contributing demographic factor rather than the sole determinant of customer value.

---

### Age Group vs Organic Buying Behaviour Analysis

```python
age_organic = df.groupby('Age Group')['Organic Buyer Flag'].mean().sort_values(ascending=False)
```

This analysis examined organic purchasing tendencies across different age groups.

The Organic Buyer Flag represents whether a customer tends to purchase organically without strong dependence on promotions, advertisements, or discounts. Since the column is binary (0 and 1), calculating the mean provides the proportion of organic buyers within each age category.

The results show that the 36–45 age group demonstrates relatively stronger organic buying behavior compared to other demographics in the dataset.

This may indicate that customers in this age category:

* have stronger brand trust,
* make more independent purchase decisions,
* rely less on promotional influence,
* exhibit mature and consistent buying behavior.

At the same time, this should be interpreted carefully. The analysis reflects trends observed in the available data and does not imply that all individuals in this demographic behave similarly.

Buying behavior is highly dependent on contextual and socioeconomic factors, including:

* household responsibilities,
* purchasing power,
* family purchasing decisions,
* cultural influences,
* product necessity,
* and economic conditions.

Therefore, the findings should be treated as demographic-level insights rather than fixed behavioral conclusions.

Customers aged 36–45 show:
*  highest customer value
*  strong organic buying behavior

This indicates:
a commercially underleveraged premium demographic.

Recommendation

Increase targeted campaigns toward mid-age premium customers while reducing blanket discounts.

In [62]:
# Question 3 — Which Age Group Spent How Much
print(df.groupby('Age Group').agg(
    Customer_Count = ('Customer ID', 'count'),
    Avg_Spend      = ('Purchase Amount (USD)', 'mean'),
    Total_Spend    = ('Purchase Amount (USD)', 'sum'),
    Avg_Value      = ('Customer Value Score', 'mean'),
    Organic_Rate   = ('Organic Buyer Flag', 'mean')
).round(2))

           Customer_Count  Avg_Spend  Total_Spend  Avg_Value  Organic_Rate
Age Group                                                                 
18-25                 502      60.74        30491    5792.36          0.57
26-35                 742      59.76        44342    5934.69          0.56
36-45                 729      59.31        43234    5829.30          0.59
46-55                 753      60.58        45619    6236.60          0.57
56+                  1105      59.06        65256    6112.72          0.57


# The Difference Between Sum and Average

| Metric       | What It Shows                                    | Problem                                      |
|--------------|--------------------------------------------------|----------------------------------------------|
| Total Spend  | How much the group contributes overall           | Bigger group always wins — not fair comparison |
| Avg Spend    | How much each customer spends on average         | Fair comparison — removes count effect       |

# Your Q3 Demographics Final Answer
>"Total spend analysis is misleading in this dataset due to unequal group sizes - the 56+ group dominates total spend at 65 dollars, 256 purely due to its size of 1,105 customers. Per-customer analysis reveals 46-55 as the highest value demographic with an average value score of 6,236 and highest average spend of $60.58, despite being a smaller group. The 36-45 segment shows the highest organic buying rate of 0.59 — meaning they require discounts least. Both 46-55 and 36-45 are commercially underlevered — strong per-customer metrics with lower brand penetration than the 56+ group."

In [63]:
"""
QUESTION 4
“How should promotions and retention strategies be restructured?”

"""
# 4.How should the brand restructure its promotional strategy to protect margins without losing volume?
"Which customers should we stop giving discounts to — and which ones still need them?"

'Which customers should we stop giving discounts to — and which ones still need them?'

In [64]:
promo = df.groupby('Discount dependency Flag').agg(
    Customer_Count  = ('Customer ID', 'count'),
    Avg_Value_Score = ('Customer Value Score', 'mean'),
    Avg_Loyalty_A   = ('Loyalty Score A', 'mean'),
    Avg_Loyalty_B   = ('Loyalty Score B', 'mean'),
    Avg_Spend       = ('Purchase Amount (USD)', 'mean'),
    Avg_Frequency   = ('Frequency Score', 'mean')
).round(2)

print(promo)

                          Customer_Count  Avg_Value_Score  Avg_Loyalty_A  \
Discount dependency Flag                                                   
0                                   2223          5910.84          29.99   
1                                   1677          6095.82          29.72   

                          Avg_Loyalty_B  Avg_Spend  Avg_Frequency  
Discount dependency Flag                                           
0                                  0.92      60.13           3.94  
1                                  1.53      59.28           3.98  


In [65]:
# Compare Value : Value of discount users vs non-discount users

discount_value = df.groupby('Discount dependency Flag')['Customer Value Score'].mean()
print(discount_value)

Discount dependency Flag
0    5910.843455
1    6095.820513
Name: Customer Value Score, dtype: float64


In [66]:
# Compare Loyalty

discount_loyalty = df.groupby('Discount dependency Flag')['Loyalty Score A'].mean()
print(discount_loyalty)

Discount dependency Flag
0    29.994152
1    29.719141
Name: Loyalty Score A, dtype: float64


In [67]:
#Compare frequency

discount_frequency = df.groupby('Discount dependency Flag')['Frequency Score'].mean()
print(discount_frequency)

Discount dependency Flag
0    3.937922
1    3.976148
Name: Frequency Score, dtype: float64


In [68]:
#Compare Purchase Amount

discount_purchase = df.groupby('Discount dependency Flag')['Purchase Amount (USD)'].mean()
print(discount_purchase)

Discount dependency Flag
0    60.130454
1    59.279070
Name: Purchase Amount (USD), dtype: float64


In [69]:
# Strategic Comparison Table

promo_analysis = pd.DataFrame({
    'Avg Customer Value': discount_value,
    'Avg Loyalty': discount_loyalty,
    'Avg Frequency': discount_frequency,
    'Avg Purchase Amount': discount_purchase
})

print(promo_analysis)

                          Avg Customer Value  Avg Loyalty  Avg Frequency  \
Discount dependency Flag                                                   
0                                5910.843455    29.994152       3.937922   
1                                6095.820513    29.719141       3.976148   

                          Avg Purchase Amount  
Discount dependency Flag                       
0                                   60.130454  
1                                   59.279070  


Interpretation

Discount-dependent customers show customer value, loyalty, and purchase frequency levels very similar to non-discount users.

The differences between both customer groups are relatively small, indicating that promotional campaigns are still contributing meaningful revenue without significantly harming retention behavior.

However, excessive dependence on blanket discounts may create long-term margin pressure.

Therefore, instead of completely removing discounts, the brand should move toward:


*   targeted promotional campaigns
*   personalized offers
*   loyalty-based rewards
*   subscription-driven retention strategies







This would help maintain customer retention while gradually reducing over-reliance on broad discounting strategies

In [70]:
df

,Customer ID,Age,Gender,Item Purchased,Category,Purchase Amount (USD),Location,Size,Color,Season,...,Previous Purchases,Payment Method,Frequency of Purchases,Discount dependency Flag,Frequency Score,Customer Value Score,Organic Buyer Flag,Loyalty Score A,Loyalty Score B,Age Group
0,1,55,Male,Blouse,Clothing,53,Kentucky,L,Gray,Winter,...,14,Venmo,Fortnightly,1,6,4452,0,20,1,46-55
1,2,19,Male,Sweater,Clothing,64,Maine,L,Maroon,Winter,...,2,Cash,Fortnightly,1,6,768,0,8,2,18-25
2,3,50,Male,Jeans,Clothing,73,Massachusetts,S,Maroon,Spring,...,23,Credit Card,Weekly,1,7,11753,0,30,2,46-55
3,4,21,Male,Sandals,Footwear,90,Rhode Island,M,Maroon,Spring,...,49,PayPal,Weekly,1,7,30870,0,56,2,18-25
4,5,45,Male,Blouse,Clothing,49,Oregon,M,Turquoise,Spring,...,31,PayPal,Annually,1,1,1519,0,32,1,36-45
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3895,3896,40,Female,Hoodie,Clothing,28,Virginia,L,Turquoise,Summer,...,32,Venmo,Weekly,0,7,6272,1,40,1,36-45
3896,3897,52,Female,Backpack,Accessories,49,Iowa,L,White,Spring,...,41,Bank Transfer,Bi-Weekly,0,5,10045,1,47,1,46-55
3897,3898,46,Female,Belt,Accessories,33,New Jersey,L,Green,Spring,...,24,Venmo,Quarterly,0,2,1584,1,27,0,46-55
3898,3899,44,Female,Shoes,Footwear,77,Minnesota,S,Brown,Summer,...,24,Venmo,Weekly,0,7,12936,1,32,1,36-45


In [71]:
# 5. 5.What does the brand's ideal customer profile look like, and how can it acquire more of them?
# 5 Key Questions to Address (directly written, numbered):

# What This Question Asks Simply

"Describe your BEST customer in data terms — specific enough that marketing can find more of them today."

'Describe your BEST customer in data terms — specific enough that marketing can find more of them today.'

In [72]:
# Step 1 — Find top 25% threshold
threshold = df['Customer Value Score'].quantile(0.75)
print('Top 25% threshold:', threshold)

# Step 2 — Filter top customers
ideal = df[df['Customer Value Score'] >= threshold]
print('Total ideal customers:', len(ideal))

Top 25% threshold: 8588.0
Total ideal customers: 975


### Missing Dimensions & Their Impact

| Missing Dimension | Problem |
|------------------|---------|
| No Demographics | Cannot target ads |
| No Behavioral | Do not know what habits to look for |
| No Commitment | Cannot identify genuine vs discount-driven |
| No Preferences | Do not know which product or state to focus on |

### Why Each Dimension Is Necessary

| Dimension     | Columns                                              | Why Necessary                                                                                                                                 |
|---------------|------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------|
| Demographics  | Age, Gender                                          | Marketing team needs this to target ads — Facebook/Instagram targeting runs on age and gender                                                  |
| Behavioral    | Frequency, Previous Purchases, Loyalty Score        | Tells brand WHAT the ideal customer does — how often, how long, how loyal                                                                     |
| Commitment    | Subscription, Organic Rate, Review Rating           | Tells brand HOW DEEPLY the ideal customer is connected — are they paying subscribers? Do they need discounts?                               |
| Preferences   | Category, Payment Method, Location                  | Tells brand WHERE to find them and WHAT they buy — which states to target, which product to use as entry point  

In [73]:
ideal = df[df['Customer Value Score'] >= df['Customer Value Score'].quantile(0.75)]

print('Total Ideal Customers:', len(ideal))
print()
print('1. Age         :', ideal['Age'].mean().round(1), 'years avg')
print('2. Gender      :', ideal['Gender'].mode()[0])
print('3. Category    :', ideal['Category'].mode()[0])
print('4. Frequency   :', ideal['Frequency of Purchases'].mode()[0])
print('5. Location    :', ideal['Location'].mode()[0])
print('6. Payment     :', ideal['Payment Method'].mode()[0])
print('7. Subscribed  :', round((ideal['Subscription Status']=='Yes').mean()*100,1), '%')
print('8. Organic     :', round((ideal['Organic Buyer Flag']==1).mean()*100,1), '%')
print('9. Avg Spend   :', ideal['Purchase Amount (USD)'].mean().round(2))
print('10. Avg Rating :', ideal['Review Rating'].mean().round(2))

Total Ideal Customers: 975

1. Age         : 44.5 years avg
2. Gender      : Male
3. Category    : Clothing
4. Frequency   : Weekly
5. Location    : New York
6. Payment     : Credit Card
7. Subscribed  : 28.5 %
8. Organic     : 56.2 %
9. Avg Spend   : 74.18
10. Avg Rating : 3.78


### The Ideal Customer — Data Backed

| # | Dimension     | Value        | Meaning                              |
|---|---------------|--------------|--------------------------------------|
| 1 | Age           | 44.5 years   | Mid-40s customer                     |
| 2 | Gender        | Male         | Male dominated segment               |
| 3 | Category      | Clothing     | Primary product preference           |
| 4 | Frequency     | Weekly       | Comes back every week                |
| 5 | Location      | New York     | Highest concentration                |
| 6 | Payment       | Credit Card  | Trust and spend capacity             |
| 7 | Subscribed    | 28.5%        | Low subscription rate                |
| 8 | Organic       | 56.2%        | More than half buy without discount  |
| 9 | Avg Spend     | $74.18       | Highest spend in dataset             |
| 10| Avg Rating    | 3.78         | Moderate satisfaction                |